File: python/day2.py
# You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

# You may assume the two numbers do not contain any leading zero, except the number 0 itself.

# Example:

# Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
# Output: 7 -> 0 -> 8
# Explanation: 342 + 465 = 807.
class ListNode:
     def __init__(self, val=0, next=None):
         self.val = val
        self.next = next

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = None
        current = None
        summa = 0
        carry = 0
        while l1 or l2:
            if l1 and l2:
                summa = carry + l1.val + l2.val
                l1 = l1.next
                l2 = l2.next
            elif l1 and not l2:
                summa = carry + l1.val
                l1 = l1.next
            elif not l1 and l2:
                summa = carry + l2.val
                l2 = l2.next
            carry = summa // 10
            if head is None:
                head = ListNode(summa % 10)
                current = head
            else:
                current.next = ListNode(summa % 10)
                current = current.next
        if carry > 0:
            current.next = ListNode(carry)
        return head

File: javascript/day2.js
//You are given two non-empty linked lists representing two non-negative integers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

//You may assume the two numbers do not contain any leading zero, except the number 0 itself.

//Example:

//Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
//Output: 7 -> 0 -> 8
//Explanation: 342 + 465 = 807.

function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */

var addTwoNumbers = function (l1, l2) {
  let carry = 0;
  let resultLN = new ListNode();
  let temp = resultLN;

  while (l1 || l2 || carry) {
    let l1_value = 0;
    let l2_value = 0;

    if (l1) l1_value = l1.val;
    if (l2) l2_value = l2.val;

    let result = l1_value + l2_value + carry;

    if (result > 9) {
      carry = 1;
    } else {
      carry = 0;
    }

    if (carry) {
      temp.val = result - 10;
    } else {
      temp.val = result;
    }

    if (carry || l1?.next || l2?.next) {
      temp.next = new ListNode();
      temp = temp.next;
      temp.next = null;
    }

    if (l1) l1 = l1.next;
    if (l2) l2 = l2.next;
  }

  return resultLN;
};
